README
NAME
    Kafka::Librd - bindings for librdkafka

SUPPORT
    Please note that this module is not actively developed and maintained,
    if you are interested in taking it over, please contact the author

SYNOPSIS
        use Kafka::Librd;

        my $kafka = Kafka::Librd->new(
            Kafka::Librd::RD_KAFKA_CONSUMER,
            {
                "group.id" => 'consumer_id',
            },
        );
        $kafka->brokers_add('server1:9092,server2:9092');
        $kafka->subscribe( \@topics );
        while (1) {
            my $msg = $kafka->consumer_poll(1000);
            if ($msg) {
                if ( $msg->err ) {
                    say "Error: ", Kafka::Librd::Error::to_string($err);
                }
                else {
                    say $msg->payload;
                }
            }
        }

DESCRIPTION
    This module provides perl bindings for librdkafka.

METHODS
  new
        $kafka = $class->new($type, \%config)

    Create a new instance. $type can be either "RD_KAFKA_CONSUMER" or
    "RD_KAFKA_PRODUCER". Config is a hash with configuration parameters as
    described in
    <https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md>,
    additionally it may include "default_topic_config" key, with a hash
    containing default topic configuration properties.

  brokers_add
        $cnt = $kafka->brokers_add($brokers)

    add one or more brokers to the list of initial bootstrap brokers.
    *$brokers* is a comma separated list of brokers in the format
    "[proto://]host[:port]".

  subscribe
        $err = $kafka->subscribe(\@topics)

    subscribe to the list of topics using balanced consumer groups.

  unsubscribe
        $err = $kafka->unsubscribe

    unsubscribe from the current subsctiption set

  subscription
        $tplist = $kafka->subscription

    return current subscription. Subscription returned as a reference to
    array of hashes with the following fields: "topic", "partition",
    "offset", "metadata".

  assign
        $err = $kafka->assign(\@tplist)

    assign partitions to consume. @tplist is an array of hashes with "topic"
    and "partition" fields set.

  assignment
        $tplist = $kafka->assignment

    return current assignment. Result returned in the same way as for
    "subscription".

  consumer_poll
        $msg = $kafka->consumer_poll($timeout_ms)

    poll for messages or events. If any message or event received, returns
    "Kafka::Librd::Message" object. If "<$msg-"err>> for returned object is
    zero (RD_KAFKA_RESP_ERR_NO_ERROR), then it is a proper message,
    otherwise it is an event or an error.

  commit
        $err = $kafka->commit(\@tplist, $async)

    commit offsets to the broker. @tplist is an array of hashes with the
    following keys: "topic", "partition", "offset", "metadata". If
    @topic_partition_list is missing or undef, then current partition
    assignment is used instead. If $async is 1, then method returns
    immediately, if it is 0 or missing then method blocks until offsets are
    commited.

  commit_message
        $err = $kafka->commit_message($msg, $async)

    commit message's offset for the message's partition. $async same as for
    "commit".

  committed
        $tplist = $kafka->committed(\@tplist, $timeout_ms)

    retrieve commited offsets for topics and partitions specified in
    @tplist, which is an array of hashes with "topic" and "partition"
    fields. Returned $tplist contains a copy of the input list with added
    "offset" fields.

  position
        $tplist = $kafka->position(\@tplist)

    retrieve current offsets for topics and partitions specified in @tplist,
    which is an array of hashes with "topic" and "partition" fields.
    Returned $tplist contains a copy of the input list with added "offset"
    fields.

  consumer_close
        $err = $kafka->consumer_close

    close down the consumer

  topic
        $topic = $kafka->topic($name, \%config)

    return a "Kafka::Librd::Topic"topic object, that can be used to produce
    messages

  outq_len
        $len = $kafka->outq_len

    return the current out queue length.

  destroy
        $kafka->destroy

    destroy kafka handle

  dump
        $kafka->dump

    dump internal state of kafka handle to stdout, only useful for debugging

Kafka::Librd::Topic
    This class maps to "rd_kafka_topic_t" structure from librdkafka and
    represents topic. It should be created with "topic" method of
    Kafka::Librd object. It provides the following method:

  produce
        $err = $topic->produce($partition, $msgflags, $payload, $key)

    produce a message for the topic. *$msgflags* can be RD_KAFKA_MSG_F_BLOCK
    in the future, but currently it should be set to 0, RD_KAFKA_MSG_F_COPY
    and RD_KAFKA_MSG_F_FREE must not be used, internally RD_KAFKA_MSG_F_COPY
    is always set.

Kafka::Librd::Message
    This class maps to "rd_kafka_message_t" structure from librdkafka and
    represents message or event. Objects of this class have the following
    methods:

  err
    return error code from the message

  topic
    return topic name

  partition
    return partition number

  offset
    return offset. Note, that the value is truncated to 32 bit if your perl
    doesn't support 64 bit integers.

  key
    return message key

  payload
    return message payload

CAVEATS
    Message offset is truncated to 32 bit if perl compiled without support
    for 64 bit integers.

SEE ALSO
    <https://github.com/edenhill/librdkafka>

BUGS
    Please report any bugs or feature requests via GitHub bug tracker at
    <http://github.com/plicease/Kafka-Librd/issues>.

AUTHOR
    Original author Pavel Shaydo "<zwon at cpan.org>"

    Current maintainer Graham Ollis "<plicease@cpan.org>"

LICENSE AND COPYRIGHT
    Copyright (C) 2016, 2017 Pavel Shaydo

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Kafka-Librd

0.18_01 2020-11-30 05:33:03 -0700
  - Fixed bug calling XSLoader (gh#37)

0.17_01 2020-11-27 16:21:52 -0700
  - Added Kafka::Librd::Error::last_error (gh#33, gh#34 harmathy++)

0.15 2020-03-15 12:28:38 -0600
  - Added support for Visual C++ + vcpkg (gh#21)

0.14 2019-10-20 19:50:39 -0600
  - Aded flush and destroy methods (gh#19, miket++)

0.12 2019-07-26 15:25:57 -0400
  - Added timestamp method (gh#10, gh#14, eserte++)

0.11 2019-07-22 09:31:39 -0400
 - Fix memory leak (eserte++ gh#13, gh#16)

0.10 2019-02-09 10:01:17 -0500
  - New maintainer Graham Ollis <plicease@cpan.org>
  - Fix compatability with Perls older than 5.8.9

0.09 Tue Oct 23 2018 Pavel Shaydo <zwon@cpan.org>
  - add note about maintenance status

0.08 Fri Mar  3 2017 Pavel Shaydo <zwon@cpan.org>
  - implement all methods from High-Level consumer API

0.07 Sat Feb  4 2017 Pavel Shaydo <zwon@cpan.org>
  - improved example scripts
  - bumped librdkafka dependency

0.06 Tue Dec  6 2016 Pavel Shaydo <zwon@cpan.org>
  - make it compile on threaded perls. See #3,
    thanks to @aero for the report

0.05 Tue Nov 29 2016 Pavel Shaydo <zwon@cpan.org>
  - minor doc fix
  - remove unnecessary Test::Most dependency. Fixes #2, thanks to
    @andk for the report.

0.04 Sat Oct 22 2016 Pavel Shaydo <zwon@cpan.org>
  - add destroy and dump methods

0.03 Mon Oct 17 2016 Pavel Shaydo <zwon@cpan.org>
  - implement producer

0.02 Fri Oct 14 2016 Pavel Shaydo <zwon@cpan.org>
  - Use Alien::Librdkafka to build librdkafka

0.01 Thu Oct 13 2016 Pavel Shaydo <zwon@cpan.org>
  - First version, released on an unsuspecting world.



Makefile.PL
use 5.008;
use strict;
use warnings;
use Config;
use ExtUtils::MakeMaker 6.64;
use Alien::Librdkafka 0.9.3;
use Alien::Base::Wrapper qw( Alien::Librdkafka !export );
use File::Spec ();

WriteMakefile(
    NAME                => 'Kafka::Librd',
    AUTHOR              => q{Graham Ollis <plicease@cpan.org>},
    LICENSE             => 'perl',
    VERSION_FROM        => 'lib/Kafka/Librd.pm',
    ABSTRACT_FROM       => 'lib/Kafka/Librd.pm',
    OBJECT              => 'Rdkafka.o rdkafkaxs.o',
    PL_FILES            => {},
    PREREQ_PM => {
        'Exporter::Lite'       => '0',
        'XSLoader'             => '0.22',
    },
    TEST_REQUIRES => {
        'Test::More'           => '0.94',
    },
    CONFIGURE_REQUIRES => {
        'Alien::Base::Wrapper' => '0',
        'Alien::Librdkafka'    => '0.9.3',
        'ExtUtils::MakeMaker'  => '6.64',
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'Kafka-Librd-* const_xs.inc' },
    depend              => { 'Rdkafka.c' => 'const_xs.inc' },
    META_MERGE => {
        resources => {
            homepage   => 'https://github.com/plicease/Kafka-Librd',
            bugtracker => 'https://github.com/plicease/Kafka-Librd/issues',
            repository => 'git://github.com/plicease/Kafka-Librd',
            license    => 'http://dev.perl.org/licenses/',
        },
        x_contributors => [
            'Pavel Shaydo <zwon@cpan.org>',
            'Graham Ollis <plicease@cpan.org>',
        ],
    },
    Alien::Base::Wrapper->mm_args,
);

sub MY::postamble {
    "const_xs.inc : @{[ File::Spec->catfile(qw( utils generate_const.pl )) ]}\n" .
    "\t\$(FULLPERL) @{[ File::Spec->catfile(qw( utils generate_const.pl )) ]}\n";
}


rdkafkaxs.c
#include "rdkafkaxs.h"
#include "ppport.h"

#define ERRSTR_SIZE 1024

rd_kafka_topic_partition_list_t*
krd_parse_topic_partition_list(pTHX_ AV* tplist) {
    char errstr[ERRSTR_SIZE];
    rd_kafka_topic_partition_list_t* tpar;

    int tplen = av_len(tplist)+1;
    tpar = rd_kafka_topic_partition_list_new(tplen);
    int i;
    for (i=0; i<tplen; i++) {
        SV** elemr = av_fetch(tplist, i, 0);
        if (elemr == NULL)
            continue;
        SV* conf = *elemr;
        if (!SvROK(conf) || strncmp(sv_reftype(SvRV(conf), 0), "HASH", 5) != 0) {
            strncpy(errstr, "elements of topic partition list expected to be hashes", ERRSTR_SIZE);
            goto CROAK;
        }
        HV* confhv = (HV*)SvRV(conf);
        SV** topicsv = hv_fetch(confhv, "topic", 5, 0);
        if (topicsv == NULL) {
            snprintf(errstr, ERRSTR_SIZE, "topic is not specified for element %d of the list", i);
            goto CROAK;
        }
        STRLEN len;
        char* topic = SvPV(*topicsv, len);
        SV** partitionsv = hv_fetch(confhv, "partition", 9, 0);
        if (partitionsv == NULL) {
            snprintf(errstr, ERRSTR_SIZE, "partition is not specified for element %d of the list", i);
            goto CROAK;
        }
        int32_t partition = SvIV(*partitionsv);
        rd_kafka_topic_partition_t* tp = rd_kafka_topic_partition_list_add(tpar, topic, partition);
        hv_iterinit(confhv);
        HE* he;
        while ((he = hv_iternext(confhv)) != NULL) {
            char* key = HePV(he, len);
            SV* val = HeVAL(he);
            if (strncmp(key, "topic", 6) == 0 || strncmp(key, "partition", 10) == 0) {
                // this we already handled
                ;
            } else if (strncmp(key, "offset", 7) == 0) {
                tp->offset = SvIV(val);
            } else if (strncmp(key, "metadata", 9) == 0) {
                tp->metadata = SvPV(val, len);
                tp->metadata_size = len;
            } else {
                snprintf(errstr, ERRSTR_SIZE, "unknown option %s for element %d of the list", key, i);
                goto CROAK;
            }
        }
    }
    return tpar;

CROAK:
    rd_kafka_topic_partition_list_destroy(tpar);
    croak("%s", errstr);
    return NULL;
}

AV* krd_expand_topic_partition_list(pTHX_ rd_kafka_topic_partition_list_t* tpar) {
    char errstr[ERRSTR_SIZE];
    AV* tplist = newAV();
    int i;
    for (i = 0; i < tpar->cnt; i++) {
        rd_kafka_topic_partition_t* elem = &(tpar->elems[i]);
        HV* tp = newHV();
        hv_stores(tp, "topic", newSVpv(elem->topic, 0));
        hv_stores(tp, "partition", newSViv(elem->partition));
        hv_stores(tp, "offset", newSViv(elem->offset));
        if(elem->metadata_size > 0) {
            hv_stores(tp, "metadata", newSVpvn(elem->metadata, elem->metadata_size));
        }
        av_push(tplist, newRV_noinc((SV*)tp));
    }
    return tplist;

CROAK:
    croak("%s", errstr);
    return NULL;
}

rd_kafka_conf_t* krd_parse_config(pTHX_ rdkafka_t *krd, HV* params) {
    char errstr[ERRSTR_SIZE];
    rd_kafka_conf_t* krdconf;
    rd_kafka_conf_res_t res;
    HE *he;

    krdconf = rd_kafka_conf_new();
    rd_kafka_conf_set_opaque(krdconf, (void *)krd);
    hv_iterinit(params);
    while ((he = hv_iternext(params)) != NULL) {
        STRLEN len;
        char* key = HePV(he, len);
        SV* val = HeVAL(he);
        if (strncmp(key, "default_topic_config", len) == 0) {
            if (!SvROK(val) || strncmp(sv_reftype(SvRV(val), 0), "HASH", 5) != 0) {
                strncpy(errstr, "default_topic_config must be a hash reference", ERRSTR_SIZE);
                goto CROAK;
            }
            rd_kafka_topic_conf_t* topconf = krd_parse_topic_config(aTHX_ (HV*)SvRV(val), errstr);
            if (topconf == NULL) goto CROAK;
            rd_kafka_conf_set_default_topic_conf(krdconf, topconf);
        } else {
            // set named configuration property
            char *strval = SvPV(val, len);
            res = rd_kafka_conf_set(
                    krdconf,
                    key,
                    strval,
                    errstr,
                    ERRSTR_SIZE);
            if (res != RD_KAFKA_CONF_OK)
                goto CROAK;
        }
    }

    return krdconf;

CROAK:
    rd_kafka_conf_destroy(krdconf);
    croak("%s", errstr);
    return NULL;
}

rd_kafka_topic_conf_t* krd_parse_topic_config(pTHX_ HV *params, char* errstr) {
    rd_kafka_topic_conf_t* topconf = rd_kafka_topic_conf_new();
    rd_kafka_conf_res_t res;
    HE *he;

    hv_iterinit(params);
    while ((he = hv_iternext(params)) != NULL) {
        STRLEN len;
        char* key = HePV(he, len);
        SV* val = HeVAL(he);
        char *strval = SvPV(val, len);
        res = rd_kafka_topic_conf_set(
                topconf,
                key,
                strval,
                errstr,
                ERRSTR_SIZE);
        if (res != RD_KAFKA_CONF_OK)
            goto ERROR;
    }

    return topconf;

ERROR:
    rd_kafka_topic_conf_destroy(topconf);
    return NULL;
}


lib/Kafka/Librd.pm
package Kafka::Librd;

use strict;
use warnings;
use XSLoader;
use Exporter::Lite;

our $VERSION = '0.20';
XSLoader::load('Kafka::Librd', $VERSION);

our @EXPORT_OK;

=head1 NAME

Kafka::Librd - bindings for librdkafka

=head1 SYNOPSIS

    use Kafka::Librd;

    my $kafka = Kafka::Librd->new(
        Kafka::Librd::RD_KAFKA_CONSUMER,
        {
            "group.id" => 'consumer_id',
        },
    );
    $kafka->brokers_add('server1:9092,server2:9092');
    $kafka->subscribe( \@topics );
    while (1) {
        my $msg = $kafka->consumer_poll(1000);
        if ($msg) {
            if ( $msg->err ) {
                say "Error: ", Kafka::Librd::Error::to_string($err);
            }
            else {
                say $msg->payload;
            }
        }
    }


=head1 DESCRIPTION

This module provides perl bindings for librdkafka.

=head1 METHODS

=cut

=head2 new

    $kafka = $class->new($type, \%config)

Create a new instance. $type can be either C<RD_KAFKA_CONSUMER> or
C<RD_KAFKA_PRODUCER>. Config is a hash with configuration parameters as
described in
L<https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md>,
additionally it may include C<default_topic_config> key, with a hash containing
default topic configuration properties.

=cut

sub new {
    my ( $class, $type, $params ) = @_;
    return _new( $type, $params );
}

{
    my $errors = Kafka::Librd::Error::rd_kafka_get_err_descs();
    no strict 'refs';
    for ( keys %$errors ) {
        *{__PACKAGE__ . "::RD_KAFKA_RESP_ERR_$_"} = eval "sub { $errors->{$_} }";
        push @EXPORT_OK, "RD_KAFKA_RESP_ERR_$_";
    }
}

=head2 brokers_add

    $cnt = $kafka->brokers_add($brokers)

add one or more brokers to the list of initial bootstrap brokers. I<$brokers>
is a comma separated list of brokers in the format C<[proto://]host[:port]>.

=head2 subscribe

    $err = $kafka->subscribe(\@topics)

subscribe to the list of topics using balanced consumer groups.

=head2 unsubscribe

    $err = $kafka->unsubscribe

unsubscribe from the current subscription set

=head2 subscription

    $tplist = $kafka->subscription

return current subscriptions. Subscription returned as a reference to array of
hashes with the following fields: C<topic>, C<partition>, C<offset>, C<metadata>.

=head2 assign

    $err = $kafka->assign(\@tplist)

assign partitions to consume. C<@tplist> is an array of hashes with
C<topic> and C<partition> fields set.

=head2 assignment

    $tplist = $kafka->assignment

return current assignment. Result returned in the same way as for
L</subscription>.

=head2 consumer_poll

    $msg = $kafka->consumer_poll($timeout_ms)

poll for messages or events. If any message or event received, returns
L</Kafka::Librd::Message> object. If C<<$msg->err>> for returned object is zero
(RD_KAFKA_RESP_ERR_NO_ERROR), then it is a proper message, otherwise it is an
event or an error.

=head2 commit

    $err = $kafka->commit(\@tplist, $async)

commit offsets to the broker. C<@tplist> is an array of hashes
with the following keys: C<topic>, C<partition>, C<offset>, C<metadata>. If
@topic_partition_list is missing or undef, then current partition assignment
is used instead. If C<$async> is 1, then method returns immediately, if it is
0 or missing then method blocks until offsets are committed.

=head2 commit_message

    $err = $kafka->commit_message($msg, $async)

commit message's offset for the message's partition. C<$async> same as for
L</commit>.

=head2 committed

    $tplist = $kafka->committed(\@tplist, $timeout_ms)

retrieve committed offsets for topics and partitions specified in C<@tplist>,
which is an array of hashes with C<topic> and C<partition> fields. Returned
C<$tplist> contains a copy of the input list with added C<offset> fields.

=head2 position

    $tplist = $kafka->position(\@tplist)

retrieve current offsets for topics and partitions specified in C<@tplist>,
which is an array of hashes with C<topic> and C<partition> fields. Returned
C<$tplist> contains a copy of the input list with added C<offset> fields.

=head2 consumer_close

    $err = $kafka->consumer_close

close down the consumer

=head2 topic

    $topic = $kafka->topic($name, \%config)

Return a L<topic|/Kafka::Librd::Topic> object, that can be used to produce
messages.

If an error occurs during creation of the topic, C<undef> is returned. In such
case use L</Kafka::Librd::Error::last_error> to obtain the corresponding error
code!

=head2 outq_len

    $len = $kafka->outq_len

return the current out queue length.

=head2 flush

    $kafka->flush($timeout_ms)

wait until all outstanding produce requests, et.al, are completed.

=head2 destroy

    $kafka->destroy

destroy kafka handle

=head2 dump

    $kafka->dump

dump internal state of kafka handle to stdout, only useful for debugging

=head1 Kafka::Librd::Topic

This class maps to C<rd_kafka_topic_t> structure from librdkafka and represents
topic. It should be created with L</topic> method of Kafka::Librd object. It
provides the following method:

=head2 produce

    $status = $topic->produce($partition, $msgflags, $payload, $key)

produce a message for the topic. I<$msgflags> can be RD_KAFKA_MSG_F_BLOCK in
the future, but currently it should be set to 0, RD_KAFKA_MSG_F_COPY and
RD_KAFKA_MSG_F_FREE must not be used, internally RD_KAFKA_MSG_F_COPY is always
set.

The returned status is -1 in case of an error, otherwise 0. The error code can
be retrieved using the L</Kafka::Librd::Error::last_error> function.

=head2 destroy

    $topic->destroy

destroy topic handle

=head1 Kafka::Librd::Message

This class maps to C<rd_kafka_message_t> structure from librdkafka and
represents message or event. Objects of this class have the following methods:

=head2 err

return error code from the message

=head2 topic

return topic name

=head2 partition

return partition number

=head2 offset

return offset. Note, that the value is truncated to 32 bit if your perl doesn't
support 64 bit integers.

=head2 key

return message key

=head2 payload

return message payload

=head2 timestamp(\$tstype)

return message timestamp (milliseconds since UNIX epoch)

The $tstype argument is optional, and if present, it should be a
scalar reference. It will be filled with one of the following values:

=over

=item

C<Kafka::Librd::RD_KAFKA_TIMESTAMP_NOT_AVAILABLE>

=item

C<Kafka::Librd::RD_KAFKA_TIMESTAMP_CREATE_TIME>

=item

C<Kafka::Librd::RD_KAFKA_TIMESTAMP_LOG_APPEND_TIME>

=back

=head1 Kafka::Librd::Error

=head2 Kafka::Librd::Error::to_string

   my $error_message =  Kafka::Librd::Error::to_string($err)

Convert an error code into a human-readable error description. Use this for
error codes returned by L</Kafka::Librd::Error::last_error> and
L<Kafka::Librd::Message::err|/err>.

=head2 Kafka::Librd::Error::last_error

    my $err = Kafka::Librd::Error::last_error

Retrieve the last error state set by function calls L</topic> and L</produce>.
This function should be called immediately after those functions, since they
store error information globally.

=cut

package Kafka::Librd::Topic;
package Kafka::Librd::Message;
package Kafka::Librd::Error;

1;

__END__

=head1 CAVEATS

Message offset is truncated to 32 bit if perl compiled without support for
64 bit integers.

=head1 SEE ALSO

L<https://github.com/edenhill/librdkafka>

=head1 BUGS

Please report any bugs or feature requests via GitHub bug tracker at
L<http://github.com/uperl/Kafka-Librd/issues>.

=head1 AUTHOR

Original author Pavel Shaydo C<< <zwon at cpan.org> >>

Current maintainer Graham Ollis C<< <plicease@cpan.org> >>

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2016, 2017 Pavel Shaydo

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut
